File: 0200-number-of-islands/0200-number-of-islands.go
type DisjointSet struct{
    parents []int
}

func NewDisjointSet(n int) DisjointSet{
    p := make([]int, n)
    for i:=0;i<n;i++{
        p[i] = -1
    }
    return DisjointSet{
        parents: p,
    }
}

func (d *DisjointSet) Find(x int) int {
    if d.parents[x] < 0 {
        return x
    }
    path := make([]int, 0)
    for d.parents[x] >= 0 {
        path = append(path, x)
        x = d.parents[x]
    }
    for _, p := range path {
        d.parents[p] = x
    }
    return x
}

func (d *DisjointSet) Union(a, b int) bool {
    if a == b {
        return false
    }

    aId, bId := d.Find(a), d.Find(b)
    if aId == bId {
        return false
    }

    aSize, bSize := d.Size(aId), d.Size(bId)
    if aSize > bSize {
        d.parents[bId] = aId
        d.parents[aId] = -aSize -bSize
    }else{
        d.parents[aId] = bId
        d.parents[bId] = -aSize -bSize
    }
    return true
}

func (d *DisjointSet) Size(a int) int {
    aId := d.Find(a)
    return -d.parents[aId]
}

func (d *DisjointSet) SetSize() int {
    c := 0
    for _, p := range d.parents {
        if p < 0 {
            c++
        }
    }
    return c
}

const (
    one = byte('1')
    zero = byte('0')
) 


func numIslands(grid [][]byte) int {
    m, n := len(grid), len(grid[0])
    set := NewDisjointSet(m*n)
    getIdx := func(row, col int) int {
        return row * n + col
    } 
    // getRC := func(idx int) (int, int) {
    //     return idx / n, idx % n
    // }
    var topIdx, leftIdx int
    var zeroIdx = -1
    var tempIdx int
    for i:=0;i<m;i++{
        for j:=0;j<n;j++{

            if grid[i][j] == zero {
                tempIdx = getIdx(i, j)

                if zeroIdx != -1 {
                    set.Union(zeroIdx, tempIdx)  
                }   
                zeroIdx = tempIdx
                // r, c := getRC(set.Find(zeroIdx))
                // assert(grid[r][c] == zero)
                continue
            }

            idx := getIdx(i, j)
            if i > 0 && grid[i-1][j] == one {
                topIdx = getIdx(i-1, j)
                set.Union(idx, topIdx)
            }

            if j > 0 && grid[i][j-1] == one {
                leftIdx = getIdx(i, j-1)
                set.Union(idx, leftIdx)
            }
            // r, c := getRC(set.Find(idx))
            // assert(grid[r][c] == one)
        }
    }

    if zeroIdx != -1 {
        return set.SetSize() - 1
    }
    return set.SetSize()
}


func assert(cond bool) {
    if !cond {
        panic("assert failed")
    }
}

File: 0200-number-of-islands/0200-number-of-islands.rs

impl Solution {
    pub fn num_islands(grid: Vec<Vec<char>>) -> i32 {
        let size = grid.len() * grid[0].len();
        let mut ds = DisjointSet::new(size as u32);
        let col_size = grid[0].len();
        let mut last_zero_id: Option<u32> = None;
        for r in 0..grid.len() {
            for c in 0..grid[r].len() {
                if grid[r][c] == '1' {
                    let curr_cell_id = to_cell_id(r, c, col_size);
                    if r >= 1 && grid[r - 1][c] == '1' {
                        ds.union(curr_cell_id, to_cell_id(r - 1, c, col_size));
                    }
                    if c >= 1 && grid[r][c - 1] == '1' {
                        ds.union(curr_cell_id, to_cell_id(r, c - 1, col_size));
                    }
                } else {
                    let curr_cell_id = to_cell_id(r, c, col_size);

                    if last_zero_id != None {
                        ds.union(curr_cell_id, last_zero_id.unwrap());
                    }
                    last_zero_id = Some(curr_cell_id);
                }
            }
        }
        let mut total = ds.total_set();
        if last_zero_id != None {
            total -= 1;
        }
        return total as _;
    }
}

fn to_cell_id(row: usize, col: usize, col_size: usize) -> u32 {
    return (row * col_size + col) as u32;
}


pub struct DisjointSet {
    parents: Vec<i32>,
}

impl DisjointSet {
    pub fn new(n: u32) -> Self {
        return DisjointSet {
            parents: vec![-1; n as usize],
        };
    }

    pub fn find(&mut self, x: u32) -> u32 {
        let mut x = x as usize;
        if self.parents[x] < 0 {
            return x as _;
        }

        let mut path = vec![];
        while self.parents[x] >= 0 {
            path.push(x);
            x = self.parents[x] as usize;
        }

        for &p in path.iter() {
            self.parents[p] = x as i32;
        }
        return x as _;
    }

    pub fn union(&mut self, a: u32, b: u32) -> bool {
        if a == b {
            return false;
        }

        let (a_id, b_id) = (self.find(a) as i32, self.find(b) as i32);

        if a_id == b_id {
            return false;
        }

        let (a_size, b_size) = (self.union_size(a_id as u32), self.union_size(b_id as u32));

        if b_size > a_size {
            self.parents[a_id as usize] = b_id as i32;
            self.parents[b_id as usize] = -((a_size + b_size) as i32);
        } else {
            self.parents[b_id as usize] = a_id as i32;
            self.parents[a_id as usize] = -((a_size + b_size) as i32);
        }
        return true;
    }

    pub fn union_size(&mut self, x: u32) -> u32 {
        let x_id = self.find(x);
        return (-self.parents[x_id as usize]) as u32;
    }

    pub fn total_set(&self) -> u32 {
        let mut result = 0;
        for &i in self.parents.iter() {
            if i < 0 {
                result += 1;
            }
        }
        return result as u32;
    }

    pub fn get_all_union_ids(&self) -> Vec<usize> {
        let mut union_ids = vec![];
        for (i, &val) in self.parents.iter().enumerate() {
            if val < 0 {
                union_ids.push(i);
            }
        }
        return union_ids;
    }
}